vms/proposervm/block/build.go

```diff
@@ -6,39 +6,152 @@ package block
 		return nil, err
 	}
 
 	return block, block.initialize(bytes)
 }
 
+func CalculateVRFOut(vrfSig []byte) [32]byte {
+	// build the hash of the following struct:
+	// +-------------------------+----------+------------+
+	// |  prefix :               | [8]byte  | "rng-derv" |
+	// +-------------------------+----------+------------+
+	// |  vrfSig :               | [96]byte |  96 bytes  |
+	// +-------------------------+----------+------------+
+	if len(vrfSig) != bls.SignatureLen {
+		return [32]byte{0}
+	}
+
+	buffer := make([]byte, len(vrfOutPrefix)+bls.SignatureLen)
+	copy(buffer, vrfOutPrefix)
+	copy(buffer[len(vrfOutPrefix):], vrfSig)
+	return hashing.ComputeHash256Array(buffer)
+}
+
+func initializeID(bytes []byte, signature []byte) (ids.ID, error) {
+	var unsignedBytes []byte
+	// The serialized form of the block is the unsignedBytes followed by the
+	// signature, which is prefixed by a uint32. So, we need to strip off the
+	// signature as well as its length prefix to get the unsigned bytes.
+	lenUnsignedBytes := len(bytes) - wrappers.IntLen - len(signature)
+
+	if lenUnsignedBytes < 0 {
+		return ids.Empty, errInvalidBlockEncodingLength
+	}
+
+	unsignedBytes = bytes[:lenUnsignedBytes]
+	return hashing.ComputeHash256Array(unsignedBytes), nil
+}
+
+// marshalBlock marshal the given statelessBlock by using either the default statelessBlock or
+// coping the exported fields into statelessBlockV0 and then marshaling it.
+// this allows the marsheler to produce encoded blocks that match the old style blocks as long as
+// the VRFSig feature was not enabled.
+func marshalBlock(block *statelessBlock[statelessUnsignedBlock]) ([]byte, error) {
+	if len(block.StatelessBlock.VRFSig) == 0 {
+		// create a backward compatible block ( without VRFSig ) and use the statelessBlockV0 encoder for the encoding.
+		var preBlockSigBlock SignedBlock = &statelessBlock[statelessUnsignedBlockV0]{
+			StatelessBlock: statelessUnsignedBlockV0{
+				ParentID:     block.StatelessBlock.ParentID,
+				Timestamp:    block.StatelessBlock.Timestamp,
+				PChainHeight: block.StatelessBlock.PChainHeight,
+				Certificate:  block.StatelessBlock.Certificate,
+				Block:        block.StatelessBlock.Block,
+			},
+			Signature: block.Signature,
+		}
+		return Codec.Marshal(CodecVersion, &preBlockSigBlock)
+	}
+	var blockIntf SignedBlock = block
+	return Codec.Marshal(CodecVersion, &blockIntf)
+}
+
+func calculateBootstrappingBlockSig(chainID ids.ID, networkID uint32) [hashing.HashLen]byte {
+	// build the hash of the following struct:
+	// +-----------------------+----------+------------+
+	// |  prefix :             | [8]byte  | "rng-root" |
+	// +-----------------------+----------+------------+
+	// |  chainID :            | [32]byte |  32 bytes  |
+	// +-----------------------+----------+------------+
+	// |  networkID:           | uint32   |  4 bytes   |
+	// +-----------------------+----------+------------+
+
+	buffer := make([]byte, len(vrfRngRootPrefix)+ids.IDLen+4)
```

Comment:
nit: could use `wrappers.IntLen` instead of `4` to remove use of a "magic number" even though it's fairly obvious from the comments here

Reply:
done.